NEW Adding a new component for "Presented options"

ScopeyNZ opened the pull request:
This component gives the content author two options for populating the "presented options" of a field (such as a dropdown field):
- Manual entry with a text area
- Choosing options from a list of "all options"

The list of all options using a CKAN API request to get distinct values from specific columns (provided as a prop to the component).

This PR also contains the integration into the existing "dropdown" filter type.

robbieaverill commented on client/src/components/PresentedOptions.js:
```diff
@@ -0,0 +1,504 @@
+   * Returns the namespaced form field name for the given field
+   *
+   * @param {string} fieldName
+   * @param {object} props If provided, will be used instead of this.props
```
This was originally so you could pass props from the constructor before they're assigned to the instance - that code is gone now so maybe we can remove this extra argument?

robbieaverill commented on client/src/components/PresentedOptions.js:
```diff
@@ -0,0 +1,504 @@
+    const loadPromises = [];
+
+    // Prep a datastore. Note that this doesn't actually fire any requests yet.
+    const Datastore = CKANApi.loadDatastore(endpoint, resource);
```
I think convention would be for this to be lowercase unless it's a React component

robbieaverill commented on client/src/components/PresentedOptions.js:
```diff
@@ -0,0 +1,504 @@
+
+    let options = [];
+    const { suggestedOptionCache } = this.state;
+    const loadPromises = [];
```
It changes below - does this need to be `let`?

ScopeyNZ replied:
If we're just mutating the object we can still use `const`. Only if we reassign the variable to a new scalar/object then we must use `let`.

robbieaverill replied:
TIL - gracias

robbieaverill commented on client/src/components/PresentedOptions.js:
```diff
@@ -0,0 +1,504 @@
+    });
+
+    // We have to wait for all promises and then just run this function again...
+    // Promise.all(loadPromises).then(() => this.loadSuggestedOptions());
```
I reckon this function could be broken up a bit - thoughts?

ScopeyNZ replied:
Yep it was getting a bit bloated. I split the `fetch` into it's own method and also the process that trims and filters the suggested options.

robbieaverill commented on client/src/components/PresentedOptions.js:
```diff
@@ -0,0 +1,504 @@
+      return null;
+    }
+
+    const description = i18n._t('.MANUAL_OPTION_DESCRIPTION', 'Options provided must match the ' +
```
We're missing the translation class name here

robbieaverill commented on client/src/components/PresentedOptions.js:
```diff
@@ -0,0 +1,504 @@
+        <InputGroup>
+          <Input value={this.state.separatorDelimiter} onChange={this.handleDelimiterChange} />
+          <InputGroupAddon addonType="append">
+            <Button onClick={this.handleExecuteSeparator} color="primary">Update</Button>
```
```suggestion
            <Button onClick={this.handleExecuteSeparator} color="primary">
              {i18n._t('CKANPresentedOptions.UPDATE', 'Update')}
            </Button>
```

robbieaverill commented on client/src/components/PresentedOptions.js:
```diff
@@ -0,0 +1,504 @@
+          </InputGroupAddon>
+        </InputGroup>
+        <FormText>
+          Split options by characters. eg. comma
```
```suggestion
          {i18n._t('CKANPresentedOptions.SPLIT_BY', 'Split options by characters. e.g. comma')}
```

robbieaverill commented on client/src/lib/CKANApi/DataStore.js:
```diff
@@ -0,0 +1,62 @@
+import CKANApi from '../CKANApi';
+
+export default class {
```
Could use a docblock

robbieaverill commented on client/src/lib/CKANApi/DataStore.js:
```diff
@@ -0,0 +1,62 @@
+    };
+
+    // If an invalid term is given then assume that it will return nothing
+    const termType = term && typeof term;
```
I'm personally a bit against this kind of ternary assignment. I read it as "termType is the [boolean] evaluation of `term && typeof term`". Maybe `const termType = term ? typeof term : null;` might be easier to read?

ScopeyNZ replied:
Fair. I will just update to early return a falsy value of `term` and then we can avoid the ternary assignment all together :)

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+        $data['data']['selectTypes'] = self::getSelectTypes();
+        $data['data']['endpoint'] = $this->getResouce()->Endpoint;
+        $data['data']['resource'] = $this->getResouce()->Identifier;
+        $data['data']['fieldMap'] = $this->getResouce()->Fields()->map('ID', 'Name')->toArray();
```
We could probably reduce the repetition of `$data['data']['` by using `array_merge()` instead

robbieaverill replied:
```suggestion
        $data['data']['fieldMap'] = $this->getResource()->Fields()->map('ID', 'Name')->toArray();
```

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+     * @param Resource $resouce
+     * @return $this
+     */
+    public function setResouce($resouce)
```
```suggestion
    public function setResource(Resource $resource)
```

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+     */
+    public function setResouce($resouce)
+    {
+        $this->resouce = $resouce;
```
```suggestion
        $this->resource = $resource;
```

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+    /**
+     * @return Resource
+     */
+    public function getResouce()
```
```suggestion
    public function getResource()
```

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+     */
+    public function getResouce()
+    {
+        return $this->resouce;
```
```suggestion
        return $this->resource;
```

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+        $data['data']['options'] = self::getOptions();
+        $data['data']['selectTypeDefault'] = self::SELECT_TYPE_DEFAULT;
+        $data['data']['selectTypes'] = self::getSelectTypes();
+        $data['data']['endpoint'] = $this->getResouce()->Endpoint;
```
```suggestion
        $data['data']['endpoint'] = $this->getResource()->Endpoint;
```

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+        $data['data']['selectTypeDefault'] = self::SELECT_TYPE_DEFAULT;
+        $data['data']['selectTypes'] = self::getSelectTypes();
+        $data['data']['endpoint'] = $this->getResouce()->Endpoint;
+        $data['data']['resource'] = $this->getResouce()->Identifier;
```
```suggestion
        $data['data']['resource'] = $this->getResource()->Identifier;
```

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+        $maxLength = null,
+        Form $form = null
+    ) {
+        $this->setResouce($resource);
```
```suggestion
        $this->setResource($resource);
```

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+     *
+     * @var Resource
+     */
+    protected $resouce;
```
```suggestion
    protected $resource;
```

robbieaverill commented on tests/Forms/PresentedOptionsFieldTest.php:
```diff
@@ -0,0 +1,18 @@
+{
+    public function testConstruct()
+    {
+        $field = new PresentedOptionsField('my-field');
```
```suggestion
        $field = new PresentedOptionsField('my-field', new Resource);
```

^ or a mocked Resource object to pass through

ScopeyNZ commented:
I've rebased and removed the WIP. I'd still like to improve test coverage (and I'll be working on that today) but I'm happy to make follow up PRs for that to prevent this PR getting any larger.

NightJar commented:
OK I did a bit more around translating the string and showing one message or the other:
![image](https://user-images.githubusercontent.com/778003/50874174-e2a11b80-1427-11e9-9194-16768f6db6ba.png)
But I'll stop working on it now until guidance comes @sachajudd :)

sachajudd commented:
[Styleguide](https://projects.invisionapp.com/dsm/silver-stripe/silver-stripe/asset/components/5bd26ef841dbd50011d50eac) @NightJar 🙂

robbieaverill commented on client/src/components/CKANPresentedOptions.js:
```diff
@@ -0,0 +1,581 @@
+ *
+ * If using a checkbox set, the user can adjust the default delimiter field.
+ */
+class CKANPresentedOptions extends Component {
```
Can you add `Field` as the component suffix to be in line with the others? #66

robbieaverill commented on client/src/components/CKANPresentedOptions.js:
```diff
@@ -0,0 +1,581 @@
+   * Given a list of options, run through and split the options by the given separator/delimiter
+   *
+   * @param options
+   * @param delimiter
```
Minor: needs types

robbieaverill commented on client/src/components/CKANPresentedOptions.js:
```diff
@@ -0,0 +1,581 @@
+  renderSeparator() {
+    return (
+      <FormGroup className="ckan-presented-options__option-separator">
+        <Label for="optionSeparator">Delimiter</Label>
```
Needs i18n

robbieaverill commented on client/src/components/CKANPresentedOptions.js:
```diff
@@ -0,0 +1,581 @@
+  // FormActionComponent: PropTypes.oneOfType([
+  //   PropTypes.string,
+  //   PropTypes.func
+  // ]).isRequired,
```
Can remove this now

robbieaverill commented on client/src/components/CKANPresentedOptions.js:
```diff
@@ -0,0 +1,581 @@
+export default fieldHolder(inject(
+  ['TextField', 'FormAction', 'Loading'],
+  (TextFieldComponent, FormActionComponent, LoadingComponent) => ({
+    TextFieldComponent,
```
This and the form action aren't needed any more, only the loading component (since we're using Reactstrap components instead)

robbieaverill commented on client/src/legacy/CKANResourceFilterForm-entwine.js:
```diff
@@ -0,0 +1,10 @@
+  $('.js-injector-boot select.ckan-columns__filter-fields').entwine({
+    onchange() {
+      const presentedOptions = this.closest('form').find('.ckan-presented-options__container');
+      presentedOptions.renderComponent(this.val());
```
Is this going to end up re-rendering the component again and again?

robbieaverill commented on src/Forms/PresentedOptionsField.php:
```diff
@@ -0,0 +1,118 @@
+            'Other',
+            'Something scrollable',
+        ];
+    }
```
Can this be removed now? I lost context about where this is used in the component since I added it 😆

NightJar commented:
@sachajudd This is how the error notice ended up looking:

![image](https://user-images.githubusercontent.com/778003/50996672-6a516c00-1587-11e9-8f9b-97d9a64ffc3e.png)

If you would like me to tidy things I've missed, please open a new issue :)

ScopeyNZ commented:
Gorgeous 👍
